### PR TITLE
Width config

### LINKED ISSUE #368: Enable configuration of relative panel width (resizable panels)
In GitLab by @rmraschkowski on Nov 26, 2021, 12:13

Some panels are regarded less important than others by stakeholders/ users. E.g. Metadata panel doesn't need as much relative width as the Text panel, especially as the transcript tab needs to show the full lines without additional linebreaks to stay true to the manuscript. Being able to configure the relative width in a config file would be nice to have. E.g. "Manuscript = 25%; Text = 30%; Commentary = 25%; Metadata = 20%

### PR DESCRIPTION
**Pull request by paulpestov**
Test with appending "width" config options to each panel config item. Check the README updates for how to configure.

Closes #368

### CODE REVIEW

File: src/components/panels/PanelsWrapper.vue [hunk 1]
```diff
@@ -1,29 +1,71 @@
...
+      class="t-flex t-h-full t-overflow-x-auto t-pb-2 lg:t-pb-0 -t-mr-4 lg:t-mr-0"
+    >
+      <Panel
+        v-for="(panel, i) in panels"
+        v-show="panel.show"
+        :key="`pc${i}`"
+        :panel="panel"
+        :active-view="getActiveView(i)"
+        :default-width="defaultWidth"
+        :class="{ 't-ml-4': i > 0 && !isMobile, 't-ml-2': i > 0 && isMobile, 't-mr-4': i === panels.length - 1 && isMobile }"
```
Comment (orlinmalkja) [inline src/components/panels/PanelsWrapper.vue:17]: a question about the syntax of conditional styling: do we apply a certain style when the part after `:` is true ?

Comment (paulpestov) [reply src/components/panels/PanelsWrapper.vue:17]: yes, exactly

File: src/utils/resize.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    onUnmounted(() => {
+      window.removeEventListener('resize', onWindowResize);
+    });
+
+    isInitialized = true;
+  }
+
+  const onResize = (callback) => {
+    resizeSubscribers.add(callback);
+    return () => resizeSubscribers.delete(callback); // Return an unsubscribe function
```
Comment (orlinmalkja) [inline src/utils/resize.js:31]: I have a question here: When will this delete function be executed ? On a normal resize of the window, it is not executed, only the function definition is returned..

File: src/components/panels/PanelsWrapper.vue [hunk 3]
```diff
@@ -1,29 +1,71 @@
...
+const container = useTemplateRef('container');
+const defaultWidth = ref(300)
+
+let unsubscribe;
+
+const updatePanelWidth = (isMobile) => {
+  if (container.value) {
+    defaultWidth.value = isMobile ? container.value.clientWidth - panelWrapperPadding : container.value.clientWidth / panels.value.filter(({ show }) => !!(show)).length - panelMargin;
+  }
+  return 300;
```
Comment (orlinmalkja) [inline src/components/panels/PanelsWrapper.vue:51]: Is this return statement needed ? I don't see whether it is assigned to some variable.

File: src/components/panels/Panel.vue [hunk 4]
```diff
@@ -193,20 +202,20 @@ export default {
...
-      () => {
-        init(props.panel.views);
-      },
-    );
+
+    watch(item,() => init(props.panel.views));
+
+    function getWidth(configValue) {
+      if (!configValue || typeof configValue !== 'number') return 1;
+      if (isMobile.value || configValue < 1) return 1;
```
[the remark this answers is not in the document]
Comment (paulpestov) [reply src/components/panels/Panel.vue:215]: To save headache with sizings below the calculated width, I decided to disallow any values below 1.

File: src/components/panels/PanelsWrapper.vue (hunk 3 above)
Comment (paulpestov) [reply src/components/panels/PanelsWrapper.vue:51]: If the container does not exist yet.

File: src/utils/resize.js (hunk 2 above)
Comment (paulpestov) [reply src/utils/resize.js:31]: here https://github.com/subugoe/tido/pull/540/files#diff-fc950e4a837261bc8651866090c3b578859d068322e29f41f88ed6fd0afbf63dR63

Comment (orlinmalkja) [reply src/utils/resize.js:31]: what I see from the above link is 'watch(configStore.config, () => {
  updatePanelWidth(isMobile.value)
})'
Is this what you meant ?

Comment (paulpestov) [reply src/utils/resize.js:31]: No, its in the PanelWrapper in the onUnount

**Review by orlinmalkja — APPROVED** [review]

**Pull request merged**